### ISSUE TITLE
mxnet homework 2-conv_layer function 

### ISSUE THREAD

**Issue by jianglong98** [BitTiger-MP/DS502-AI-Engineer#22]
Comparing to mlp_layer, I need help in defining conv_layer.
As Chuck said in the class, we can add convolution/activation/pooling layers in "def conv_layer", but I found the similar function in mlp is realized in "def get_mlp_sym()", not in "def mlp_layer".
So my question is what we should input in "def conv_layer".

I attached the origin mlp_layer and my get_conv_sym() below for reference.

def mlp_layer(input_layer, n_hidden, activation=None, BN=False):
    # get a FC layer
    l = mx.sym.FullyConnected(data=input_layer, num_hidden=n_hidden)
    # get activation, it can be relu, sigmoid, tanh, softrelu or none
    if activation is not None:
        l = mx.sym.Activation(data=l, act_type=activation)
    if BN:
        l = mx.sym.BatchNorm(l)
    return l

def get_conv_sym():     // I created by myself since here
    """
    :return: symbol of a convolutional neural network
    """
    data = mx.sym.Variable("data")
    # todo: design the CNN architecture
    # How deep the network do you want? like 4 or 5
    # How wide the network do you want? like 32/64/128 kernels per layer
    # How is the convolution like? Normal CNN? Inception Module? VGG like?
    data_f = mx.sym.flatten(data=data)

    l = mx.sym.Convolution(data=data_f, kernel=(3, 3), num_filter=20)
    l = mx.sym.Activation(data=l, act_type="relu")
    l = mx.sym.Pooling(data=l, pool_type="max", kernel=(2, 2), stride=(2, 2))
    # first fullc
    l = mx.sym.FullyConnected(data=l, num_hidden=10)
    # softmax loss
    l = mx.sym.SoftmaxOutput(data=l, name='softmax')
    return l

**Comment by ChuckTan123** [comment]
@jianglong98 
Thank you for the great question. 
If I understood your question correctly,  you are asking if we need to implement pool (or any other function) either in the "def conv_layer" or directly in "def get_conv_symbol". 

It is really based on you design to make yourself less coding. 
Think about this, you have 1000 conv layer, you prefer calling "conv_layer" 1000 times, OR you write 1000 time "mx.sym.Convolution" AND "mx.sym.Activation" AND "mx.sym.Pooling"?

The only reason that we want to have mlp_layer with a lot of if-else option, is that we don't want to repeat ourselves. 

So for the design of "def conv_layer", you can put convolution, activation, pool, BN as optional controlled by input parameters. 

Or you can also repeat yourself by copying and pasting the same code (pooling) in "def get_conv_symbol", as there are just few lines of the codes.

Hope it helps.

**Comment by guan4015** [comment]
[Week 2 Homework](https://github.com/guan4015/DS502_AI_bittiger/tree/master/Chuck/mxnet-week2)

This is Xiao's homework.
